Description: --with-openssl-legacy-provider of PHP 8.4: load the OpenSSL 3 legacy provider (RC4, Blowfish, DES, RC2, CAST5, MD4 and others) together with the default provider at startup; off by default, no change with OpenSSL 1.1.
Origin: php-src PHP-8.4 ext/openssl/config0.m4 and openssl.c (GH-13951, "ext/openssl: Add option to load legacy algorithm provider"), the same code

diff --git a/ext/openssl/openssl.c b/ext/openssl/openssl.c
--- a/ext/openssl/openssl.c
+++ b/ext/openssl/openssl.c
@@ -59,6 +59,7 @@
 #if PHP_OPENSSL_API_VERSION >= 0x30000
 #include <openssl/core_names.h>
 #include <openssl/param_build.h>
+#include <openssl/provider.h>
 #endif
 
 #if (OPENSSL_VERSION_NUMBER < 0x10100000L || defined(LIBRESSL_VERSION_NUMBER)) && !defined(OPENSSL_NO_ENGINE)
@@ -1277,6 +1278,10 @@
 	OpenSSL_add_all_algorithms();
 	SSL_load_error_strings();
 #else
+#if PHP_OPENSSL_API_VERSION >= 0x30000 && defined(LOAD_OPENSSL_LEGACY_PROVIDER)
+	OSSL_PROVIDER_load(NULL, "legacy");
+	OSSL_PROVIDER_load(NULL, "default");
+#endif
 	OPENSSL_init_ssl(OPENSSL_INIT_LOAD_CONFIG, NULL);
 #endif
 
diff --git a/ext/openssl/config0.m4 b/ext/openssl/config0.m4
--- a/ext/openssl/config0.m4
+++ b/ext/openssl/config0.m4
@@ -14,6 +14,13 @@
   [whether to use system default cipher list instead of hardcoded value],
   [AS_HELP_STRING([--with-system-ciphers],
     [OPENSSL: Use system default cipher list instead of hardcoded value])],
+  [no],
+  [no])
+
+PHP_ARG_WITH([openssl-legacy-provider],
+  [whether to load legacy algorithm provider],
+  [AS_HELP_STRING([--with-openssl-legacy-provider],
+    [OPENSSL: Load legacy algorithm provider in addition to default provider])],
   [no],
   [no])
 
@@ -40,4 +47,7 @@
   if test "$PHP_SYSTEM_CIPHERS" != "no"; then
     AC_DEFINE(USE_OPENSSL_SYSTEM_CIPHERS,1,[ Use system default cipher list instead of hardcoded value ])
   fi
+  if test "$PHP_OPENSSL_LEGACY_PROVIDER" != "no"; then
+    AC_DEFINE(LOAD_OPENSSL_LEGACY_PROVIDER,1,[ Load the OpenSSL legacy algorithm provider in addition to the default provider ])
+  fi
 fi
